Register unregistered refs and add a test to check registration

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80497

Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

davidberard98 committed:
Register "clamp" decomposition

Added a missing `@register_decomposition` which will register the op
into `torch._decomp.decomposition_table`.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register "clamp" decomposition"


Added a missing `register_decomposition` which will register the op
into `torch._decomp.decomposition_table`.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

ezyang commented:
Thanks, the test for registration is much needed

ezyang commented on test/test_ops.py:
```diff
@@ -1519,12 +1519,82 @@ class TestRefsOpsInfo(TestCase):
+        '_refs.masked_fill',
+        '_refs.transpose',
+        '_refs.var',
+        # these are not aten ops?
```
Yeah, most of these are actually implemented in Python (e.g., tanhshrink). Actually, this poses a code duplication problem that we should discuss with @mruberry and @ngimel, because in many cases the original implementations of these functions in PyTorch proper would have worked just fine, should we really have two equivalent implementations of a function

ezyang commented on test/test_ops.py:
```diff
@@ -1519,12 +1519,82 @@ class TestRefsOpsInfo(TestCase):
+        '_refs.zeros',  # missing "layout"
+        '_refs.zeros_like',  # missing "layout"
+        # not sure
+        '_refs.copy_to',
```
@mruberry should this really be a ref, there's no `torch.copy_to` in the main API

ezyang commented on test/test_ops.py:
```diff
@@ -1519,12 +1519,82 @@ class TestRefsOpsInfo(TestCase):
+        # not sure
+        '_refs.copy_to',
+        '_refs.clone',
+        '_refs.equal',
```
would be good to know what's going on with these two

davidberard98 replied:
Added comments for these

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]

ezyang commented on test/test_ops.py:
```diff
@@ -1519,12 +1519,83 @@ class TestRefsOpsInfo(TestCase):
+        '_refs.as_strided',  # _prims._as_strided_meta: "reduce() of empty sequence with no initial value"
+        '_refs.copy_to',  # torch._C._jit_get_operation: No such operator aten::copy_to
+        '_refs.clone',  # test_meta.py: view size is not compatible with input tensor's size and stride
+        '_refs.equal',  # 'bool' object has no attribute 'dtype'
```
sounds like the ref doesn't implement scalar inputs correctly

davidberard98 committed:
Update on "Register unregistered refs and add a test to check registration"


Added missing `register_decomposition`s which will register the refs so
they can be used for decompositions.

Also added a test for verifying that new refs are registered.

[ghstack-poisoned]
